src/test/java/com/google/devtools/build/lib/actions/ActionCacheCheckerTest.java

```diff
@@ -391,6 +432,135 @@ public void testDeletedConstantMetadataOutputCausesReexecution() throws Exceptio
         .isNotNull();
   }
 
+  private RemoteFileArtifactValue createRemoteMetadata(String content) {
+    byte[] bytes = content.getBytes(UTF_8);
+    return new RemoteFileArtifactValue(digest(bytes), bytes.length, 1, "action-id");
+  }
+
+  @Test
+  public void saveOutputMetadata_remoteFileMetadataSaved() throws Exception {
+    // arrange
+    enableSaveOutputMetadata();
+
+    Artifact.DerivedArtifact output =
+        new Artifact.DerivedArtifact(
+            artifactRoot, PathFragment.create("bin/dummy"), NULL_ARTIFACT_OWNER);
+    String content = "content";
+    Action action = new InjectOutputMetadataAction(output, createRemoteMetadata(content));
+
+    // act
+
+    // Not cached.
+    runAction(action);
+
+    // assert
+    assertThat(output.getPath().exists()).isFalse();
+    ActionCache.Entry entry = cache.get(output.getExecPathString());
+    assertThat(entry).isNotNull();
+    assertThat(entry.getOutputFile(output)).isEqualTo(createRemoteMetadata(content));
+    assertStatistics(0, new MissDetailsBuilder().set(MissReason.NOT_CACHED, 1).build());
+  }
+
+  @Test
+  public void saveOutputMetadata_localFileMetadataNotSaved() throws Exception {
```

Comment:
Should we also test an action which injects remote metadata and creates the file locally. Think running remote runner with the option to materialize all outputs.

Reply:
Added test for the case where the local metadata is same as injected remote one.

Reply:
I was also thinking about the case when an action both injects the remote metadata and stores a file locally. Wouldn't remote execution with download all do that?

Reply:
Remote execution with download all won't inject remote metadata. but I will probably change it to inject remote metadata to support async download. Test case added.

Reply:
FYI, that is what we do internally.